elide-core/src/main/java/com/yahoo/elide/graphql/PersistentResourceFetcher.java

```diff
@@ -263,12 +266,382 @@ private Object fetchRelationship(Environment context) {
+    /**
+     * handle UPSERT operation
+     * @param context environment encapsulating graphQL's request environment
+     * @return list of {@link PersistentResource} objects
+     */
+    private Set<PersistentResource> upsertObjects(Environment context) {
+        /* sanity check for id and data argument w UPSERT */
+        if(context.ids.isPresent()) {
+            throw new BadRequestException("UPSERT must not include ids");
+        }
 
-    private Object upsertObjects(Environment context) { return null; }
+        if(!context.data.isPresent()) {
+            throw new BadRequestException("UPSERT must include data argument");
+        }
 
-    private Object removeObjects(Environment context) { return null; }
+        EntityDictionary dictionary = context.requestScope.getDictionary();
 
-    private Object deleteObjects(Environment context) { return null; }
-}
+        Set upsertOps = new HashSet();
+        for(Map<String, Object> input : context.data.get()) {
+            String idFieldName = context.isRoot() ?
+                    dictionary.getIdFieldName(dictionary.getEntityClass(context.field.getName())) :
+                    dictionary.getIdFieldName(context.parentResource.getResourceClass());
+
+            /* fetch id value fed to 'data' argument */
+            Optional<String> id = input.entrySet().stream()
+                    .filter(entry -> idFieldName.equalsIgnoreCase(entry.getKey()))
+                    .map(e -> (String)e.getValue())
+                    .findFirst();
+
+            upsertOps.addAll(inputObjectToUpsertOrUpdateOperations(input, id, context.parentResource,
+                    context.requestScope, context.field.getName(), context.outputType));
+        }
+
+        return upsertOps;
+    }
+
+    /**
+     * utility class to support storing a triplet on each node of the queue used while upserting
+     */
+    private class Triplet {
+        private List<Map<String, Object>> input;
+        private PersistentResource parentResource;
+        private String fieldName;
+
+        /**
+         * class constructor
+         * @param input data input
+         * @param parentResource parent
+         * @param fieldName field name
+         */
+        private Triplet(List<Map<String, Object>> input, PersistentResource parentResource, String fieldName) {
+            this.input = input;
+            this.parentResource = parentResource;
+            this.fieldName = fieldName;
+        }
+    }
+
+    /**
+     * handles both updating/creating by first handling the attributes and then the relationships, queuing up
+     * any further nested relationships
+     * @param input data input
+     * @param parent parent
+     * @param requestScope request scope
+     * @param fieldName field name
+     * @return set of persistent resource objects
+     */
+    private Set<PersistentResource> inputObjectToUpsertOrUpdateOperations(Map<String, Object> input, Optional<String> id,
```

Comment:
They cyclomatic complexity of this function is 9.  It might make sense to break this up a bit.